chore: add PR text in new file just to be able to create a PR

ganlhi opened the pull request:
Technical and architectural choices
===================================

Technologies used
-----------------

For this project, ReactJS 16 is used, along with React Router. The components are implemented as function components and use React Hooks to provide state management and side effects.

The correctness of data structures passed between components is ensured by PropTypes.

The UI is built with [@heetch/flamingo-react](https://github.com/heetch/flamingo).

Another possible choice I considered for types correctness is TypeScript, as I am used to work with this technology. But in this case I chose to leave it aside because I have never used it with React and I was concerned it would slow me down too much to perform the assignment in the time allowed. Plus it was a good opportunity to try to model as much as I could with PropTypes.

Data workflow in the application
--------------------------------

The central piece of data the application uses is the response of the `permissions` API. It is used to provide two types of information:

-   The structure of the multi-level navigation, and all the URLs the user can navigate to
-   The data needed to perform API queries in each page displayed to the user (read, add and update operations)

Therefore, my first concern was to process this API response to extract these two information and format them in a way they would be easy to use for the intended purposes.

This meant building a recursive array of navigation items, which is passed to the `Sidebar` component to build the application menu. This array is ordered as instructed, and contains only the visible items.

It also meant building a map between URL slugs and the actions permitted: for each slug, the object containing the URLs and HTTP methods for each API operation is stored, then it is retrieved by the corresponding page (`ProductsPage` and `CitiesPage`) to perform the requests.

As these data are mandatory to do anything meaningful in the application, only a loading spinner is shown to the user as long as it is not available.

Code architecture
-----------------

After a first "naive" version of each part of the application (products and cities), I refactored the code to extract common pieces into more generic components. The structure and basic logic of these two pages is very similar, the only difference being what information is displayed in the list of items. So I created the `WithActions` component which handles all the logic regarding the following:

-   checking the user is allowed to see the current page
-   displaying a loading spinner while the data is loading
-   displaying an error message if the loading fails
-   delegating the rendering of the list to a specific component
-   handling the add/edit workflow, as the same data is editable in both cases (name and description)

For this last part, a second generic component is used: `EditPanel`. It is implemented as a side panel, visible when creating a new item or updating an existing one. The form it contains is the same in both cases, and displays a spinner while the data is sent to the API and an error message if the write operation fails.

Testing
-------

I didn't have the time to cover everything I wanted with automated tests, mainly because I was lacking experience with the testing library for React components. That said, I tried to highlight some important aspects to be tested:

-   The utilities functions in `lib/api.js`, mainly the ones performing some complex processing. In particular the processing of the response of the `permissions` API, described in *Data workflow in the application*.
-   The different *loading state*, *normal state* and *error state* of several components.
-   The consequences of non permitted actions on the UI, like not showing the *Add* button if creating a new item is not permitted.

I would also have liked to test the consequences of succeeding and failing requests to the API, using a tool like [jest-fetch-mock](https://www.npmjs.com/package/jest-fetch-mock).

Future improvements
-------------------

The error handling is pretty basic here, and a lot could be done to provide a better user experience. For instance, we could implement auto-retry of requests failing for non explicit reasons (network errors, timeouts, etc...).

The initial loading of `permissions`, blocking the application while it is not done, could greatly benefit from a caching strategy, maybe using the ServiceWorker.

In terms of UX, the lengthy multi-level menu for continents and countries could be replaced by a search feature based on autocomplete fields. A map used to select the country by clicking on it could also be considered.

ganlhi committed:
chore: add PR text in new file just to be able to create a PR

tamarasaurus commented:
Hi @ganlhi, thanks for putting in the time to complete this technical test and for your detailed explanation of your code. Overall we thought that your implementation fulfilled the requirements and you cleanly implemented the UI components and API handling. We have some questions for you and some specific comments about parts of your code:

- What was your motivation to choose Create React App to handle the build? Do you have experience with other bundlers like Webpack or Parcel?
- We liked the way you organised the shared PropTypes in a single file (inspired by Typescript? 😉)
- You had Prettier set up in the repository - did it come with CRA? Do you normally work with any linters? (We noticed some unused variables that a linter could have caught).
- Thank you for the comprehensive documentation with JSDoc, it made it simple to understand the structure of your code 🙂
- For your `Sidebar` component, why did you choose to abstract it into a separate file?
- In the app you relied only on Flamingo for the UI - do you already have experience with other styling libraries? E.g. any CSS frameworks, styled components, emotion etc..
- Did you consider developing the UI in a responsive way if you had more time? What would you do differently?
- For your `<Route />` components, was there a reason why you hardcoded the paths here?

[https://github.com/ganlhi/heetch-tech-test/blob/43402b1c8bf7d708f06b1ea7aa6c536eb8aac4a9/src/App.js#L47-L52](https://github.com/ganlhi/heetch-tech-test/blob/43402b1c8bf7d708f06b1ea7aa6c536eb8aac4a9/src/App.js#L47-L52)

- We thought that the abstraction of the data fetching and sorting inside `api.js` well organised and showed a clear separation of concerns
- We noticed that your tests are mainly based around checking the rendering of elements and data and so the coverage was not very comprehensive. You had imported `fireEvent` in your `WithAction` test - did you intend to test the interactions?
    - Do you have experience with any end-to-end test libraries?
- For the permissions handling - when there is no action key returned for an entity (cities/products) you'll have the possibility of the code breaking here:

[https://github.com/ganlhi/heetch-tech-test/blob/43402b1c8bf7d708f06b1ea7aa6c536eb8aac4a9/src/lib/api.js#L24-L27](https://github.com/ganlhi/heetch-tech-test/blob/43402b1c8bf7d708f06b1ea7aa6c536eb8aac4a9/src/lib/api.js#L24-L27)

- We thought that you did a great job with the error and permissions handling inside `WithActions` 👏 The error retry and the generic `renderList` prop were a nice touch.
- Regarding the `EditPanel` component, how would you handle backend validation errors for specific fields?
    - Do you already have experience with any form state management libraries?
- In your 'Future Improvements' section you mentioned improving the error handling, could you elaborate on that?

Thanks again and looking forward to hearing your answers :slightly_smiling_face:

ganlhi commented:
Hi @tamarasaurus, thank you for your comments and questions. I'll try to answer all of them as best as I can. 

> What was your motivation to choose Create React App to handle the build? Do you have experience with other bundlers like Webpack or Parcel?

As there was no technical prerequisites for the test I used CRA as a simple solution to bootstrap the project, as it provides out of the box tooling for hot reloading, bundling and development web server. I do have some experience with Webpack (although I did not use the last versions, so there might be some novelties I know little about). In particular I built a Webpack config from scratch for an AngularJS project I worked on a few years ago. I also used Parcel to setup some personal experimental projects, mainly with Javascript + Elm.

> We liked the way you organised the shared PropTypes in a single file (inspired by Typescript? :wink:)

Yes, I do have this habit coming from my daily TypeScript usage. I usually define all business models as TS interfaces in a central location for a given feature module, in the Angular projects I work on. It makes it easy to model a domain even before starting implementing components. I tried to replicate this pattern with PropTypes.

> You had Prettier set up in the repository - did it come with CRA? Do you normally work with any linters? (We noticed some unused variables that a linter could have caught).

I added Prettier myself, as I am used to work with it. I think it frees the mind from trivial things like checking for consistent indentation, lines length, trailing commas, etc... It also ensure a whole team commits code with a consistent style, making readability and maintenance easier.

Regarding the linter, I usually use tslint as it is provided by default by the Angular CLI. To be honest I did not notice CRA did not provide eslint. I previously have used eslint and jshint. 

> Thank you for the comprehensive documentation with JSDoc, it made it simple to understand the structure of your code :slightly_smiling_face:

Thank you. I find it helpful, for future understanding of the code, to clearly describe the public API of the components.

> For your `Sidebar` component, why did you choose to abstract it into a separate file?

After I implemented the API functions I made a quick mockup (with pen and paper) of what the finished app would look like, and a split it into components, not being very clear yet how each one of them would be implemented. I did not thought the Sidebar would be so simple, and in the end I didn't take the time to consider a refactor. 

> In the app you relied only on Flamingo for the UI - do you already have experience with other styling libraries? E.g. any CSS frameworks, styled components, emotion etc...

I do have some experience with various libraries, not especially in the React ecosystem, as my React experience was limited. I worked with CSS frameworks like Bootstrap (who did not?) and Materialize, with components libraries like Angular Material, NG Prime and MD Bootstrap. I also have a lot of experience with SASS/SCSS. 
Regarding CSS-in-js technologies like styled components or emotion, I don't have concrete experience with them, although I kept myself informed about how they work and what are their benefits. 

> Did you consider developing the UI in a responsive way if you had more time? What would you do differently?

If I had more time yes, I would have considered it. I would change things mainly on two subjects:

First I would have the general layout respond to the viewport width, mainly the sidebar being replaced either by a sliding menu accessible through a burger icon button or by a top screen navigation. 

Then I would probably choose different components in some cases, more suitable to a mobile and touch screen experience. The main menu for instance, being a tree of links, does not seem very practical nor very readable on a mobile screen. 

> For your `<Route />` components, was there a reason why you hardcoded the paths here?

There was no particular reason other than that this was a very small and simple router setup and I didn't feel the need to add another level of abstraction when I wrote it.

> We thought that the abstraction of the data fetching and sorting inside `api.js` well organised and showed a clear separation of concerns

I like to abstract data fetching, business rules and data transformation like that because I prefer to expose to the components an API that says what can be done, not how it is done. It allows, among other things, to change the way we communicate with the backend without having to refactor the components. 

> We noticed that your tests are mainly based around checking the rendering of elements and data and so the coverage was not very comprehensive. You had imported `fireEvent` in your `WithAction` test - did you intend to test the interactions?

I lacked time to cover more ground with the tests. I would have liked to test user interactions with `fireEvent`, and to mock HTTP calls to test normal and error cases of the `api.js`'s functions. 

> Do you have experience with any end-to-end test libraries?

I do have some experience with Protractor, Cypress and TestCafe. Not on my current project, but on previous ones.

> For the permissions handling - when there is no action key returned for an entity (cities/products) you'll have the possibility of the code breaking here

You are absolutely right. I should have written:

```javascript
return sortPermissions(permissions).reduce(
    (nav, { name, item }) => {
      if (!item.actions || item.actions.show !== true) return nav;
```

Obviously my tests don't cover this case...

> We thought that you did a great job with the error and permissions handling inside `WithActions` :clap: The error retry and the generic `renderList` prop were a nice touch.

Thank you. I first wrote `CitiesPage` and `ProductsPage` as two independent components but when I saw the similarities I thought it was a good opportunity to refactor while adding some genericity. 

> Regarding the `EditPanel` component, how would you handle backend validation errors for specific fields?

I would replace the boolean `savingError` by an object describing the error for each field, and use this object in `EditPanel` to display error messages next to each relevant field. It would also be nice to disable the save button while some of these field with errors are not modified by the user. 

> Do you already have experience with any form state management libraries?

Not in the React ecosystem. I do always use the official Angular's `ReactiveFormsModule` on a daily basis, which provides form structure modelization, standard validators, ability to define custom validators, automatic injection of CSS classes based on form field status... 

On the React side, even if I never used them, I read a bit about Formik and React Hook Forms. 

> In your 'Future Improvements' section you mentioned improving the error handling, could you elaborate on that?

In its current state, the error handling is pretty basic and only consists of displaying an error instead of the expected content, and at most offer a retry option. I already mentioned improvements that can be made on errors occuring when saving the edit form, but in the case of loading lists data, for instance, the errors are unlikely to be caused by the client. Which means these are probably caused by a non-responding server, timeouts... In these cases, display a generic error message to the user and letting him deal with it is a bad user experience. We could instead automatically with an incremental delay (to avoid overloading the server). We could also notify the user that the data could not be refreshed and display cached data instead (if any).

I hope I answered all your questions and thank you for the feedback :)
